lib/expression_parser.rb

```diff
@@ -50,7 +50,18 @@ def write_vm_code(vm_writer, symbol_table)
     expression_list.each do |expression|
       expression.write_vm_code(vm_writer, symbol_table)
     end
-    vm_writer.write_call("#{class_name}.#{subroutine_name}", expression_list.length)
+
+    name = class_name
+    arg_length = expression_list.length
+    kind = symbol_table.kind_of(class_name)
+    if kind != :NONE
+      name = symbol_table.type_of(class_name)
```

Comment:
Right! So if the “receiver” is a variable whose type we know, we use the name of its type to decide which function to call.

It’s worth noting that this makes Jack’s method calls _monomorphic_: just by reading the code, we can tell exactly which function will be called by `game.run()` (or whatever) because we know that `game` is a variable of type `SquareGame` (or whatever). This is different from more complex OO languages like Ruby or Java, where the actual class of the receiver — and therefore which method implementation to call — is often not known until runtime:

```ruby
class Foo
  def greeting
    'hi'
  end
end

class Bar
  def greeting
    'yo'
  end
end

greeter_class = [Foo, Bar].sample # pick one at random
greeter = greeter_class.new
puts greeter.greeting # what does this print?
```

This means that Jack is giving us _some_ of the advantages of OO programming but also missing some pretty major ones. It would be possible for Jack to support polymorphism, but it’s obviously simpler that it doesn’t.